rspace/src/main/scala/coop/rchain/rspace/LockFreeInMemoryStore.scala

```diff
@@ -0,0 +1,253 @@
+    extends IStore[C, P, A, K]
+    with CloseOps {
+
+  protected[rspace] type Transaction = InMemTransaction[State[C, P, A, K]]
+
+  override private[rspace] def createTxnRead(): Transaction = {
+    failIfClosed()
+    new NoopTxn[State[C, P, A, K]]
+  }
+  override private[rspace] def createTxnWrite(): Transaction = {
+    failIfClosed()
+    new NoopTxn[State[C, P, A, K]]
+  }
+  override private[rspace] def withTxn[R](txn: Transaction)(f: Transaction => R) = f(txn)
+  override def close(): Unit                                                     = super.close()
+
+  type TrieTransaction = T
+
+  private implicit val codecC: Codec[C] = sc.toCodec
+  private implicit val codecP: Codec[P] = sp.toCodec
+  private implicit val codecA: Codec[A] = sa.toCodec
+  private implicit val codecK: Codec[K] = sk.toCodec
+
+  private val stateGNAT: TrieMap[Blake2b256Hash, GNAT[C, P, A, K]] =
+    TrieMap[Blake2b256Hash, GNAT[C, P, A, K]]()
+  private val stateJoin: TrieMap[C, Seq[Seq[C]]] = TrieMap[C, Seq[Seq[C]]]()
+
+  private[this] val refine       = Map("path" -> "inmem")
+  private[this] val entriesGauge = Kamon.gauge("entries").refine(refine)
+
+  private[rspace] def updateGauges() =
+    entriesGauge.set(stateGNAT.readOnlySnapshot.size.toLong)
+
+  private[rspace] def hashChannels(channels: Seq[C]): Blake2b256Hash =
+    StableHashProvider.hash(channels)
+
+  override def withTrieTxn[R](txn: Transaction)(f: TrieTransaction => R): R =
+    trieStore.withTxn(trieStore.createTxnWrite()) { ttxn =>
+      f(ttxn)
+    }
+
+  private[rspace] def getChannels(txn: Transaction, key: Blake2b256Hash): Seq[C] =
+    stateGNAT.get(key).map(_.channels).getOrElse(Seq.empty)
+
+  private[rspace] def getData(txn: Transaction, channels: Seq[C]): Seq[Datum[A]] =
+    stateGNAT.get(hashChannels(channels)).map(_.data).getOrElse(Seq.empty)
+
+  private[this] def getMutableWaitingContinuation(
+      txn: Transaction,
+      channels: Seq[C]
+  ): Seq[WaitingContinuation[P, K]] =
+    stateGNAT
+      .get(hashChannels(channels))
+      .map(_.wks)
+      .getOrElse(Seq.empty)
+
+  private[rspace] def getWaitingContinuation(
+      txn: Transaction,
+      channels: Seq[C]
+  ): Seq[WaitingContinuation[P, K]] =
+    getMutableWaitingContinuation(txn, channels)
+      .map { wk =>
+        wk.copy(continuation = InMemoryStore.roundTrip(wk.continuation))
+      }
+
+  def getPatterns(txn: Transaction, channels: Seq[C]): Seq[Seq[P]] =
+    getMutableWaitingContinuation(txn, channels).map(_.patterns)
+
+  private[rspace] def getJoin(txn: Transaction, channel: C): Seq[Seq[C]] =
+    stateJoin.getOrElse(channel, Seq.empty)
+
+  private[rspace] def joinMap: Map[Blake2b256Hash, Seq[Seq[C]]] =
+    stateJoin.map {
+      case (k, v) => (Blake2b256Hash.create(Codec[C].encode(k).map(_.toByteArray).get), v)
+    }.toMap
+
+  private[rspace] def putDatum(txn: Transaction, channels: Seq[C], datum: Datum[A]): Unit = {
+    val hash = hashChannels(channels)
+    val v = stateGNAT
```

Comment:
As I see here (maybe too paranoid?): if two threads will call `putDatum` on the same channel concurrently, one of them will overwrite result of the other, if the execution sequence will be: `[Thrd1] stateGNAT.get`; `[Thrd2] stateGNAT.get`; `[Thrd1] stateGNAT.put`; `[Thrd2] stateGNAT.put`

Reply:
From chat with @dzajkowski :  LockFree implementation very tightly bound to the FineGrainedLocking RSpace which takes care of keeping per channel locks.